FXyz-Core/src/main/java/org/fxyz3d/shapes/primitives/TorusMesh.java

```diff
@@ -1,353 +1,341 @@
+                        (float) getRadius(),
+                        (float) getTubeRadius(),
+                        (float) getTubeStartAngleOffset(),
+                        (float) getxOffset(),
+                        (float) getyOffset(),
+                        (float) getzOffset()
+                )
+        );
+    }
+    
+    private TriangleMesh createTorusMesh(
+            final int radiusDivisions,
+            final int tubeDivisions,
+            final float radius,
+            final float tRadius,
+            final float tubeStartAngle,
+            final float xOffset,
+            final float yOffset,
+            final float zOffset) {
+
+        final int numVerts = tubeDivisions * radiusDivisions;
+        int faceCount = numVerts * 2;
+
+        float[] points = new float[numVerts * 3];
+        float[] texCoords = new float[numVerts * 2];
+        int[] faces = new int[faceCount * 6];
+
+        int pointIndex = 0;
+        int texIndex = 0;
+        int faceIndex = 0;
+
+        float tubeFraction = 1.0f / tubeDivisions;
+        float radiusFraction = 1.0f / radiusDivisions;
+
+        float TWO_PI = 2.0f * 3.141592653589793f;
+
+        // create points
+        // create texCoords
+        for (int tubeIndex = 0; tubeIndex < tubeDivisions; tubeIndex++) {
+
+            float radian = tubeStartAngle + tubeFraction * tubeIndex * TWO_PI;
+
+            for (int radiusIndex = 0; radiusIndex < radiusDivisions; radiusIndex++) {
+
+                float localRadian = radiusFraction * radiusIndex * TWO_PI;
+
+                points[pointIndex + 0] = (radius + tRadius * ((float) cos(radian))) * ((float) cos(localRadian) + xOffset);
+                points[pointIndex + 1] = (radius + tRadius * ((float) cos(radian))) * ((float) sin(localRadian) + yOffset);
+                points[pointIndex + 2] = (tRadius * (float) sin(radian) * zOffset);
+
+                pointIndex += 3;
+
+                float r = radiusIndex < tubeDivisions ? tubeFraction * radiusIndex * TWO_PI : 0.0f;
+                texCoords[texIndex + 0] = (float) (sin(r) * 0.5) + 0.5f;
+                texCoords[texIndex + 1] = (float) (cos(r) * 0.5) + 0.5f;
+
+                texIndex += 2;
+            }
+        }
+
+        //create faces        
+        for (int point = 0; point < tubeDivisions; point++) {
+            for (int crossSection = 0; crossSection < radiusDivisions; crossSection++) {
+
+                final int p0 = point * radiusDivisions + crossSection;
+
+                int p1 = p0 >= 0 ? p0 + 1 : p0 - radiusDivisions;
+                p1 = p1 % radiusDivisions != 0 ? p0 + 1 : p0 + 1 - radiusDivisions;
+
+                final int p0r = p0 + radiusDivisions;
+
+                final int p2 = p0r < numVerts ? p0r : p0r - numVerts;
+
+                int p3 = p2 < (numVerts - 1) ? p2 + 1 : p2 + 1 - numVerts;
+                p3 = p3 % radiusDivisions != 0 ? p2 + 1 : p2 + 1 - radiusDivisions;
+
+                try {
```

Comment:
What is this `try-catch` for?

Reply:
I don't know. I suspect index out of bounds. I didn't want to change it in this commit, but it will have to go.

Reply:
Well, can the code reach a situation where an AIOOB can happen even? If so, it means that we need to do something about it in the `catch` block.

From what I see, the size of `faces` is `faceCount * 6 = (numVerts * 2) * 6 = (tubeDivisions * radiusDivisions)  * 12`. The loop runs over `tubeDivisions`  and `radiusDivisions`, and each iteration increments `faceIndex` by 12, so I don't think we can get out of bounds.

I would also move `int faceIndex = 0;` to just before the loops where it is used, but we can do it in another pass.

Reply:
Yes, that was my plan for the future pass -- to see if AIOOB can happen.